### PR TITLE
Design Patterns: Codec Factory

### PR DESCRIPTION
**Pull request by denisdimitrov**
CodecFactory which creates JSONCodec and XMLCodec with 2 implementations. The first one is with "new" and then second one is using Reflection.

### CODE REVIEW

**Comment by denisdimitrov** [comment]
Reminder, review this when you are available.

File: src/main/java/codecFactory/Codec.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package codecFactory;
+
+/**
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public interface Codec {
```
Comment (denisdimitrov) [inline src/main/java/codecFactory/Codec.java]: Ok.

File: src/main/java/codecFactory/Codec.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+    void getCodec();
```
Comment (denisdimitrov) [inline src/main/java/codecFactory/Codec.java]: I will fix it in a minute.

**Comment by denisdimitrov** [comment]
Removed duplication and added documentation.

File: src/main/java/codecFactory/CodecFactory.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+/**
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public interface CodecFactory {
+    /**
+     * Interface with addCodec method() for adding/creating codec.
+     * @return
+     */
+    Codec addCodec();
```
Comment (EmilGeorgiev) [inline src/main/java/codecFactory/CodecFactory.java]: We have method ```addCodec```, but we have not parameters to it. I expect when read method which begin with ```add``` to have at least on parameter. Also Codec interface have method ```get``` but is return ```void```. This is smell.

Comment (denisdimitrov) [inline src/main/java/codecFactory/CodecFactory.java]: Should I change the get method to show? It's going to sound better for void.

**Comment by denisdimitrov** [comment]
I renamed get() to show() and I also added parameter in the addCodec() method.

File: src/main/java/codecFactory/JSONCodecFactory.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+/**
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public class JSONCodecFactory implements CodecFactory {
+    /**
+     * Adds JSONCodec using "new"
+     * @return
+     */
+    public Codec addCodec(Codec codec) {
```
Comment (mgenov) [inline src/main/java/codecFactory/JSONCodecFactory.java]: The role of Factory is to create things, not to add them.

File: src/main/java/codecFactory/JSONReflectionCodec.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public class JSONReflectionCodec implements CodecFactory {
+    /**
+     * Adds and returns JSONCodec using reflection. And throws InstantiationException and IllegalAccessException
+     * if it catches them it prints out the stackTrace.
+     * @return
+     */
+    public Codec addCodec(Codec codec) {
+        Class json = new JSONCodecFactory().getClass();
```
Comment (mgenov) [inline src/main/java/codecFactory/JSONReflectionCodec.java]: The role of the Factory is to create objects but you are passing the object that need to be created as parameter ?

If you know the type of the object that need to be created then you don't need to use Reflection

```java
Codec codec = factory.createCodec("JSON"); 
```

Codec addCodec(Codec codec) looks like misleading method signature that shouldn't accept codec.

File: src/main/java/codecFactory/CodecFactory.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+/**
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public interface CodecFactory {
+    /**
+     * Interface with addCodec method() for adding/creating codec.
+     * @return
+     */
+    Codec addCodec(Codec codec);
```
Comment (mgenov) [inline src/main/java/codecFactory/CodecFactory.java]: The role of the Factory is to create things. Why your factory accepts the thing that need to be created as parameter?

**Comment by denisdimitrov** [comment]
Codecs are now being held in a Map.

**Comment by denisdimitrov** [comment]
I added two more scenarios in the tests.

**Pull request merged**